app/src/org/commcare/CommCareHeartbeatManager.java

```diff
@@ -0,0 +1,309 @@
+
+    private final HttpResponseProcessor responseProcessor = new HttpResponseProcessor() {
+
+        @Override
+        public void processSuccess(int responseCode, InputStream responseData) {
+            try {
+                String responseAsString = StreamsUtil.inputStreamToByteArray(responseData).toString();
+                JSONObject jsonResponse = new JSONObject(responseAsString);
+                parseHeartbeatResponse(jsonResponse);
+            }
+            catch (JSONException e) {
+                Logger.log(AndroidLogger.TYPE_ERROR_SERVER_COMMS,
+                        "Heartbeat response was not properly-formed JSON: " + e.getMessage());
+            }
+            catch (IOException e) {
+                Logger.log(AndroidLogger.TYPE_ERROR_SERVER_COMMS,
+                        "IO error while processing heartbeat response: " + e.getMessage());
+            }
+        }
+
+        @Override
+        public void processRedirection(int responseCode) {
+            processErrorResponse(responseCode);
+        }
+
+        @Override
+        public void processClientError(int responseCode) {
+            processErrorResponse(responseCode);
+        }
+
+        @Override
+        public void processServerError(int responseCode) {
+            processErrorResponse(responseCode);
+        }
+
+        @Override
+        public void processOther(int responseCode) {
+            processErrorResponse(responseCode);
+        }
+
+        @Override
+        public void handleIOException(IOException exception) {
+            Logger.log(AndroidLogger.TYPE_ERROR_SERVER_COMMS,
+                    "Encountered IOException while getting response stream for heartbeat response: "
+                            + exception.getMessage());
+        }
+
+        private void processErrorResponse(int responseCode) {
+            Logger.log(AndroidLogger.TYPE_ERROR_SERVER_COMMS,
+                    "Received error response from heartbeat request: " + responseCode);
+        }
+    };
+
+    private static CommCareHeartbeatManager INSTANCE;
+    public static CommCareHeartbeatManager instance() {
+        if (INSTANCE == null) {
+            INSTANCE = new CommCareHeartbeatManager();
+        }
+        return INSTANCE;
+    }
+
+    public void startHeartbeatCommunications() {
+        TimerTask heartbeatTimerTask = new TimerTask() {
+            @Override
+            public void run() {
+                try {
+                    User currentUser = CommCareApplication.instance().getSession().getLoggedInUser();
+                    //simulateRequestGettingStuck();
+                    //requestHeartbeat(currentUser);
+                    parseTestHeartbeatResponse();
+                } catch (SessionUnavailableException e) {
+                    // Means the session has ended, so we should stop these requests
+                    stopHeartbeatCommunications();
+                }
+            }
+        };
+
+        heartbeatTimer = new Timer();
```

Comment:
Addressed by always canceling the existing Timer, if it’s non-null, before instantiating a new one